### ISSUE TITLE
Continue running ZStream after an error

### ISSUE THREAD

**Issue by n3ziniuka5** [zio/zio#7349]
It would be great if there was a way to keep the stream running after an error is encountered. Use-case: consume a message from Pulsar/Kafka, etc. and pass it through multiple processing stages, then finally either positively or negatively acknowledge the message. Various processing stages can fail for a variety of reasons. So if, for example, deserialization from json fails, I don't want to kill the whole stream, I just want to handle errors for that element while others continue processing. It doesn't make sense to restart the whole stream with in-flight elements in various processing stages just for a single failure.

`.either.mapZIO` or `.mapZIO(onSuccess).tapError(onFailure)` would work great here if it didn't stop the stream on first failure.

```scala
ZStream(...)
  .via(parse)
  .via(...)
  .via(...)
  .either
  .mapZIO {
    case Right(_) => // positive acknowledgement
    case Left(error) => // log error and negative acknowledgement
  } 
```
So then it doesn't matter in which stage an error happens, it's possible to handle all of them at the very end. At the moment I am pushing an Either through the whole stream, but it's not very productive because you have to write custom versions of `mapZIO` and other operators to work nicely with `Either`. And if I handle errors in each processing stage separately, then I have to push an `Option[A]` throughout the stream which has the same problem.

ZIO's error channel would be ideal for this problem and it would be great if stream elements would be continuously routed to either the error channel or the success channel.

EDIT:
I gave an example with `tapError`, but I think that one should kill the stream on failure, because it doesn't actually catch the errors and doesn't set error channel's type to `Nothing`. So there should be an operator that allows you to handle the errors and changes the error channel to `Nothing`, indicating that the stream cannot fail and will run until a success value is produced (or indefinitely)

**Comment by vigoo** [comment]
Aren’t you looking for `catchAll`?

**Comment by adamgfraser** [comment]
The contract of a stream is that it produces zero or more elements and terminates with either exactly one error or exactly one done value. If you want to have element level errors then you need to use either an `Either` or handle element level errors within a stage, for example by logging them and dropping them. So for example instead of:

```scala
val stream1: ZStream[Any, Nothing, Int] =
  ZStream(1, 2, 3, 4, 5)

val stream2: ZStream[Any, String, Int] =
  stream1.mapZIO { n =>
    if (n % 2 == 0) ZIO.succeed(n * 3) else ZIO.fail(s"$n was not even")
  }
// terminates on first failure
```

Do this:

```scala
val stream1: ZStream[Any, Nothing, Int] =
  ZStream(1, 2, 3, 4, 5)

val stream2: ZStream[Any, Nothing, Either[String, Int]] =
  stream1.map { n =>
    if (n % 2 == 0) Right(n * 3) else Left("not even")
  }
// propagates element level errors for handling further downstream
```

Or:

```scala
val stream1: ZStream[Any, Nothing, Int] =
  ZStream(1, 2, 3, 4, 5)

val stream2: ZStream[Any, Nothing, Int] =
  stream1.mapZIO { n =>
    if (n % 2 == 0) ZIO.succeed(Some(n * 3))
    else ZIO.logError(s"$n was not even").as(None)
  }.collectSome
// handles element level errors immediately by logging and dropping them
```

**Comment by n3ziniuka5** [comment]
> Aren’t you looking for `catchAll`?

For `catchAll` you need to provide an alternative stream. And if that one can fail as well, it must also have an alternative stream. You cannot simply ignore/handle the errors and continue.

Thanks for the response @adamgfraser. Indeed, I am currently using `Either` for this problem, but will also try `Option` like you suggested because `.collectSome` can make it a bit easier to work with than with `Either`.

I understand that streams are working as intended and this is merely a suggestion. I would argue, that the current contract optimizes for non realistic scenarios, in practical applications the failures are everywhere and restarting the stream is too costly.  We can work with `Option` or `Either`, but that sacrifices the great ergonomics that we enjoy with ZIO everywhere else.

That being said, I also understand that my opinion is subjective and your experience might be different from mine, but I hope you will consider this as a potential improvement for the future. It might be possible to support this even without breaking the current contract. If you have `ZStream[R, Throwable, A]`, then you know that running it will either terminate in `Throwable` or `A`. But if you handle the errors with let's say `.catchAllContinue` and you end up with `ZStream[R, Nothing, A]`, then you know that the only way the stream terminates is with value `A`.

**Comment by adamgfraser** [comment]
The issue is fairly fundamental. When a stream is done with an error that's it. There is nothing more to run after that.

You are right that you could have an alternative design where there were "element level" errors. However, it seems like you also need "stream level" errors (e.g. you're reading from a file and the file isn't found, or you're parsing a large data structure and get an unexpected input).  So then you have two possibilities.

One possibility is that you collapse "element level" and "stream level" errors into one type parameter, in which case the semantics of a stream become extremely unclear. Currently a `ZStream[R, E, A]` means "zero or more elements of type `A` and then either terminates with done value or with an error of type `E`. In this world that description would be something like "zero or more elements of either type `A` or type `E` and then either done or possibly one further value of type `E`. I think this would also not make it possible to handle stream versus element level errors in a sane way.

The other possibility would be to add a separate type parameter for element level errors. This would certainly be possible, but having two error channels would I think have an unacceptable impact on ergonomics.

Another consideration is that the concept of an error being "terminal" seems to be built into many other streaming frameworks, such as reactive streams. Using a different model would have a significant impact on the ability to interoperate, and I think the fact that it is a very widely used framework despite having "stream level" errors reflects the nature of the underlying domain.

**Comment by n3ziniuka5** [comment]
Yes, I agree. Previously I worked with Akka streams and got used to [supervision-strategies](https://doc.akka.io/docs/akka/current/stream/stream-error.html#supervision-strategies), they are quite nice for the use-case, except that adding cleanup effects in deciders is cumbersome.

It's probably possible to implement `catchAllZIO`:
```scala
val s1 = ZStream(1, 2, 3) ++ ZStream.dieMessage("Oh! Boom!") ++ ZStream(4, 5)

val stream = s1.catchAllZIO(_ => ZIO.succeed(9))
```

For it to be consistent with how current `catchAll` works, it would have to output `1, 2, 3, 9`. But I'd like it to output `1, 2, 3, 9, 4, 5`. I completely agree that such behavior wouldn't fit into the current `ZStream`. And at that point there isn't really a need to implement `catchAllZIO` because you can just use `catchAll(_ => ZStream.fromZIO(9))`.

However, perhaps in the future it would be nice to have different types of streams. For example, `ZStream.repeatZIO` could return, let's say `RepeatedZStream[R, E, A]`. The idea is that with `repeatZIO` you want the stream to endlessly repeat an effect and emit a value, the type indicates clearly that it's an infinite stream. Then, `catchAllZIO` fits quite nicely - the infinite stream consists of elements produced by `.repeatZIO(...)`, but if it errors out, then the element is instead produced by `catchAllZIO`. If `catchAllZIO` itself fails, then the whole stream terminates with that failure. Otherwise, `ZStream.repeatZIO(...)` continues producing the elements because the failure was handled. But if there aren't more requests for similar behavior then it's obviously not worth the effort.

Thanks for your time @adamgfraser.
